### ISSUE TITLE
Just how to activate minibatching

### ISSUE THREAD

**Issue by suwangcompling** [clab/dynet#653]
New to dynet, rookie question: if I want to run dynet in Jupyter Notebook, just how exactly do I activate minibatching. 
Some example code is appreciated. Didn't get much out of ``just add --dynet autobatch in command line''. 

Thanks.

**Comment by suwangcompling** [comment]
PS: learning by making simple mult-classifiers with DyNet vs. Tensorflow, the following is my code. DyNet runs at least 100 times longer than the TF code, but not sure where my configs are not right.

`TF code:

import tensorflow as tf

from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("/tmp/data/", one_hot=True)
print "Train sizes", mnist.train.images.shape, mnist.train.labels.shape
print "Test sizes", mnist.test.images.shape, mnist.test.labels.shape
print

def reset_graph():
    if 'sess' in globals() and sess: sess.close()
    tf.reset_default_graph()
    
import random

reset_graph()

X_ph = tf.placeholder(tf.float32, shape=[None, 784])
y_ph = tf.placeholder(tf.float32, shape=[None, 10])

W = tf.Variable(tf.truncated_normal(shape=[784, 10], stddev=0.01))
b = tf.Variable(tf.zeros([10]))

pred = tf.nn.softmax(tf.matmul(X_ph, W) + b)
loss = tf.reduce_mean(-tf.reduce_sum(y_ph*tf.log(pred), reduction_indices=1))

trainer = tf.train.AdamOptimizer().minimize(loss)

num_epochs = 1
batch_size = 10
count = 0
print "Train"
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for _ in xrange(num_epochs):
        for _ in range(int(mnist.train.num_examples/batch_size)):
            count += 1
            X_batch, y_batch = mnist.train.next_batch(batch_size)
            _, loss_ = sess.run([trainer, loss], feed_dict={X_ph:X_batch,y_ph:y_batch})
            if count % 500 == 0:
                print '... loss at', str(count)+':', loss_
    
    print 
    print "Test"
    correct_pred = tf.equal(tf.argmax(pred, 1), tf.argmax(y_ph, 1))
    accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))
    print "Accuracy:", accuracy.eval({X_ph:mnist.test.images, y_ph:mnist.test.labels})`

`DyNet code

from dynet import *
dp = DynetParams()
dp.set_autobatch(True)
dp.set_mem(2048)

renew_cg()

m = ParameterCollection()
pW = m.add_parameters((10, 784))
pb = m.add_parameters(10)

trainer = AdamTrainer(m)

W = parameter(pW)
b = parameter(pb)
x = inputVector(np.zeros(784))
def calculate_loss(X_i, y_i):
    x.set(X_i)
    pred = softmax((W*x)+b)
    loss = -log(pick(pred, np.argmax(y_i)))
    return loss
count = 0
num_epochs = 1
batch_size = 10
for _ in xrange(num_epochs):
    for _ in range(int(mnist.train.num_examples/batch_size)):
        count += 1
        X_batch, y_batch = mnist.train.next_batch(batch_size)
        losses = []
        for X_i,y_i in zip(X_batch,y_batch):
            loss = calculate_loss(X_i,y_i)
            losses.append(loss)
        loss = esum(losses)
        if count % 500 == 0:
            print '... loss at', str(count)+':', loss.value()
        loss.forward()
        loss.backward()
        trainer.update()`

**Comment by pmichel31415** [comment]
Hi, in order to setup the global parameters programmatically, you need to import `_dynet` or `_gdynet` for gpu as mentionned here: http://dynet.readthedocs.io/en/latest/python.html#using-the-gpu-from-python

However the doc is not very clear about this, i'll mark this ans an enahncement

**Comment by yoavg** [comment]
@suwangcompling re your speed diff from TF, I noticed that you call `renew_cg()` only once. You should be calling it after each batch (after each call to `.backward()` in your case) and follow it by 
```
W = parameter(pW)
b = parameter(pb)
```
to add them to the new graph.

As it is now, your graph grows with each new training example.

**Comment by neubig** [comment]
@pmichel31415 I was thinking of fixing the doc, but realized I actually wasn't sure how. Is there a good example that I could work with?

**Comment by pmichel31415** [comment]
I think maybe the best thing to do would be to emhasize the `import _dynet/import dynet` difference in the  documentation of the `DynetParams` directly

**Comment by pmichel31415** [comment]
OK I've updated the docs of DynetParams: http://dynet.readthedocs.io/en/latest/python_ref.html#dynet.DynetParams

I'll consider this fixed
